Fix incorrect ORCA plan for  "Left Semi Join with replicated outer table"

24nishant opened the pull request:
**Problem**
---------------------------------------------------- 
ORCA generated incorrect plan while doing "Left Semi Join with replicated outer table" . Duplicated rows were observed in the output.


**Wrong Output:**
----------------------------------------------------
 SETUP
```
    create table repli_t1 (a int) distributed replicated;
    insert into repli_t1 values(1);
    analyze repli_t1;

    create table dist_t1 (a int , b int) distributed by (a);
    insert into dist_t1 select i, 1 from generate_series(1, 100) i;
    analyze dist_t1;
```

Query -
    `select * from repli_t1 where exists ( select 1 from dist_t1 where repli_t1.a >= dist_t1.b);`

```
Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1324036.44 rows=1 width=4) (actual time=30.539..30.665 rows=3 loops=1)
  ->  GroupAggregate  (cost=0.00..1324036.44 rows=1 width=4) (actual time=11.077..11.080 rows=2 loops=1)
        Group Key: repli_t1.a, repli_t1.ctid, repli_t1.gp_segment_id
        ->  Sort  (cost=0.00..1324036.44 rows=1 width=14) (actual time=11.062..11.063 rows=2 loops=1)
              Sort Key: repli_t1.ctid, repli_t1.gp_segment_id
              Sort Method:  quicksort  Memory: 75kB
              Executor Memory: 178kB  Segments: 3  Max: 60kB (segment 0)
              ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..1324036.44 rows=1 width=14) (actual time=9.566..9.569 rows=2 loops=1)
                    Hash Key: repli_t1.ctid, repli_t1.gp_segment_id
                    ->  GroupAggregate  (cost=0.00..1324036.44 rows=1 width=14) (actual time=2.417..2.419 rows=1 loops=1)
                          Group Key: repli_t1.a, repli_t1.ctid, repli_t1.gp_segment_id
                          ->  Sort  (cost=0.00..1324036.44 rows=12 width=14) (actual time=2.357..2.375 rows=38 loops=1)
                                Sort Key: repli_t1.ctid, repli_t1.gp_segment_id
                                Sort Method:  quicksort  Memory: 80kB
                                Executor Memory: 178kB  Segments: 3  Max: 60kB (segment 0)
                                ->  Nested Loop  (cost=0.00..1324036.44 rows=12 width=14) (actual time=1.122..1.229 rows=38 loops=1)
                                      Join Filter: (repli_t1.a >= dist_t1.b)
                                      ->  Seq Scan on repli_t1  (cost=0.00..431.00 rows=1 width=14) (actual time=0.180..0.181 rows=1 loops=1)
                                      ->  Materialize  (cost=0.00..431.00 rows=34 width=4) (actual time=0.464..0.502 rows=20 loops=2)
                                            ->  Seq Scan on dist_t1  (cost=0.00..431.00 rows=34 width=4) (actual time=0.914..0.938 rows=38 loops=1)
Optimizer: GPORCA
Planning Time: 96.918 ms
  (slice0)    Executor memory: 60K bytes.
  (slice1)    Executor memory: 106K bytes avg x 3 workers, 106K bytes max (seg0).  Work_mem: 60K bytes max.
  (slice2)    Executor memory: 60K bytes avg x 3 workers, 60K bytes max (seg0).  Work_mem: 60K bytes max.
Memory used:  128000kB
Execution Time: 78.246 ms
(27 rows)

```
select * from repli_t1 where exists ( select 1 from dist_t1 where repli_t1.a >= dist_t1.b);
	 a
	---
 	1
 	1
 	1
	(3 rows)


**RCA**
-------------------------------------------------- 
The problem was found in transform "CXformLeftSemiJoin2InnerJoinUnderGb".

Here, "Left Semi Join" is tranformed into "Inner join under aggregate". But, for a Left Semi Join with replicated table is not always equivalent to an Inner Join. For eg - if join condition is satisfied at multiple segments.

In SemiJoin, we need one row from the outer table, even if multiple rows in the inner table satisfies the join condition. With Replicated outer table, inner join can cause duplication of data, as same data will be generated from all the segments.

SAMPLE RUN 

<img width="340" alt="image" src="https://github.com/greenplum-db/gpdb/assets/99574850/28c29c4e-34bc-448c-8e0a-a6d1eb82762a">




Following the above plan, even after deduplication, Redistribution and again deduplication, the gather motion- 3:1 
will cause duplicated result 




**Fix**
------------------------------------------------
To fix the problem, CXformLeftSemiJoin2InnerJoinUnderGb transform is not performed when the outer table is replicated.

    --Correct Output
```
           Result  (cost=0.00..1324032.70 rows=1 width=4)
               Filter: (SubPlan 1)
               ->  Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.00 rows=1 width=4)
                     ->  Seq Scan on repli_t1  (cost=0.00..431.00 rows=3 width=4)
               SubPlan 1
                 ->  Result  (cost=0.00..431.00 rows=2 width=4)
                       Filter: (repli_t1.a >= dist_t1.b)
                       ->  Materialize  (cost=0.00..431.00 rows=5 width=8)
                             ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..431.00 rows=5 width=8)
                                   ->  Seq Scan on dist_t1  (cost=0.00..431.00 rows=2 width=8)
             Optimizer: GPORCA
            (11 rows)

```
	select * from repli_t1 where exists ( select 1 from dist_t1 where repli_t1.a >= dist_t1.b);
      	a
      	---
      	1
     	(1 row)


**Other Consideration** 
------------------------------------------------
Since at the time of transformation (CXformLeftSemiJoin2InnerJoinUnderGb) we do not have information about the distribution spec of the Outer child (as it depend on the physical implementation), **should  this transform be disabled by default?**

pobbatihari reviewed (approved):
Nice test coverage. Verified changes locally, LGTM

gpopt commented:
We should derive NULL key collection for replicated table if there is no primary or unique key. In other words, (ctid, gp_segment_id) should not be used as the unique keys for replicated table.
In addition, we need to add test cases, where the replicated table has primary / unique keys.

wuyuhao28 commented:
After this fix, ORCA's plan will be a single QE join plan, I think there may be some performance issue in executor, is it possible that ORCA can produce planner's plan with _RowIdExpr_ :
```
explain select * from repli_t t1 where exists ( select 1 from dist_t t2 where t1.a >= t2.b);
                                                    QUERY PLAN
-------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=2.84..2.94 rows=7 width=4)
   ->  Result  (cost=2.84..2.85 rows=2 width=4)
         ->  Unique  (cost=2.84..2.85 rows=2 width=4)
               Group Key: (RowIdExpr)
               ->  Sort  (cost=2.84..2.84 rows=2 width=4)
                     Sort Key: (RowIdExpr)
                     ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..2.82 rows=2 width=4)
                           Hash Key: (RowIdExpr)
                           ->  Nested Loop  (cost=0.00..2.78 rows=2 width=4)
                                 Join Filter: (t1.a >= t2.b)
                                 ->  Broadcast Motion 1:3  (slice3; segments: 1)  (cost=0.00..1.03 rows=1 width=4)
                                       ->  Seq Scan on repli_t t1  (cost=0.00..1.01 rows=1 width=4)
                                 ->  Seq Scan on dist_t t2  (cost=0.00..1.33 rows=33 width=4)
 Optimizer: Postgres-based planner
(14 rows)
```

vraghavan78 commented:
> 's plan will be a single QE join plan, I think ther

Planner planner broadcasts the replicated table. Orca does not. More thought needs to happen to fix the performance issue. But first step to remove wrong results.

24nishant commented:
> We should derive NULL key collection for replicated table if there is no primary or unique key. In other words, (ctid, gp_segment_id) should not be used as the unique keys for replicated table.
> In addition, we need to add test cases, where the replicated table has primary / unique keys.

Yes, I have updated the code with these.
1.  Deriving 'Null' key set for replicated table, with no explicit key
2. Added test cases for replicated table with primary key

gpopt commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -2683,9 +2682,12 @@ CTranslatorRelcacheToDXL::RetrieveRelKeysets(CMemoryPool *mp, OID oid,
+	// 2. Skip addition of {segid, ctid} as a key for replicated table,
+	// as same data is present across segments thus seg_id,
+	// will not help in defining a unique tuple.
+	if ((should_add_default_keys) &&
```
() is not needed here

24nishant replied:
Corrected.

24nishant commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -2683,9 +2682,12 @@ CTranslatorRelcacheToDXL::RetrieveRelKeysets(CMemoryPool *mp, OID oid,
+	// as same data is present across segments thus seg_id,
+	// will not help in defining a unique tuple.
+	if ((should_add_default_keys) &&
+		(rel_distr_policy != IMDRelation::EreldistrReplicated))
```
Corrected.

gpopt reviewed (commented):
I was wondering what if we run the following query with optimizer on and off:
`select distinct ctid, gp_segment_id from replicated_table`?

24nishant commented:
> select distinct ctid, gp_segment_id from replicated_table

In both the case we get 'error' that the column (ctid/gp_segment_id) does not exist. 

```
nishants4=# set optimizer to on;
SET
nishants4=#
nishants4=# select distinct ctid, gp_segment_id  from repli_t1;
ERROR:  column "ctid" does not exist
LINE 1: select distinct ctid, gp_segment_id  from repli_t1;
                        ^
nishants4=# select distinct gp_segment_id, ctid  from repli_t1;
ERROR:  column "gp_segment_id" does not exist
LINE 1: select distinct gp_segment_id, ctid  from repli_t1;
                        ^
nishants4=# set optimizer to on;
SET
nishants4=# select distinct ctid, gp_segment_id  from repli_t1;
ERROR:  column "ctid" does not exist
LINE 1: select distinct ctid, gp_segment_id  from repli_t1;
                        ^
nishants4=# select distinct gp_segment_id, ctid  from repli_t1;
ERROR:  column "gp_segment_id" does not exist
LINE 1: select distinct gp_segment_id, ctid  from repli_t1;
``` 

I also referred "[ Greenplum Create Table description](https://docs.vmware.com/en/VMware-Greenplum/7/greenplum-database/ref_guide-sql_commands-CREATE_TABLE.html)" about replicated table, found following : 
"The hidden system columns (ctid, cmin, cmax, xmin, xmax, and gp_segment_id) cannot be referenced in user queries on replicated tables because they have no single, unambiguous value. Greenplum Database returns a column does not exist error for the query."

24nishant commented:
> select distinct ctid, gp_segment_id from replicated_table

In both the case we get 'error' that the column (ctid/gp_segment_id) does not exist. 

```
nishants4=# set optimizer to on;
SET
nishants4=# select distinct ctid, gp_segment_id  from repli_t1;
ERROR:  column "ctid" does not exist
LINE 1: select distinct ctid, gp_segment_id  from repli_t1;
                        ^
nishants4=# select distinct gp_segment_id, ctid  from repli_t1;
ERROR:  column "gp_segment_id" does not exist
LINE 1: select distinct gp_segment_id, ctid  from repli_t1;
                        ^
nishants4=# set optimizer to off;
SET
nishants4=# select distinct ctid, gp_segment_id  from repli_t1;
ERROR:  column "ctid" does not exist
LINE 1: select distinct ctid, gp_segment_id  from repli_t1;
                        ^
nishants4=# select distinct gp_segment_id, ctid  from repli_t1;
ERROR:  column "gp_segment_id" does not exist
LINE 1: select distinct gp_segment_id, ctid  from repli_t1;
```

I also referred "[Greenplum Create Table](https://docs.vmware.com/en/VMware-Greenplum/7/greenplum-database/ref_guide-sql_commands-CREATE_TABLE.html)" description for 'replicated tables, found the following : 
"The hidden system columns (ctid, cmin, cmax, xmin, xmax, and gp_segment_id) cannot be referenced in user queries on replicated tables because they have no single, unambiguous value. Greenplum Database returns a column does not exist error for the query."

24nishant commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -2683,9 +2682,12 @@ CTranslatorRelcacheToDXL::RetrieveRelKeysets(CMemoryPool *mp, OID oid,
+	// 2. Skip addition of {segid, ctid} as a key for replicated table,
+	// as same data is present across segments thus seg_id,
+	// will not help in defining a unique tuple.
+	if (should_add_default_keys &&
```
Yes, Corrected.

gpopt reviewed (commented):
Hmm, I was thinking maybe we should just stop saving (ctid, gp_segment_id) as keys for replicated table into the catalog table. In that case, we don't need to update Orca's code at all.

24nishant commented:
> Hmm, I was thinking maybe we should just stop saving (ctid, gp_segment_id) as keys for replicated table into the catalog table. In that case, we don't need to update Orca's code at all.

Please correct me if I am wrong :

I checked, we cannot explicitly create a primary key constriant using ctid or gp_segment_id. 
So, this entry cannot go in pg_constraint and we are not required to stop saving it. For internal use, we added 'ctid, gp_segment_id' as a key for all the tables. This addition we intend to stop through the PR.

Query Output --
```
nishants4=# create table repli_t_pk (a int, PRIMARY KEY(a, ctid)) distributed replicated;
ERROR:  cannot alter system column "ctid"

nishants4=# create table repli_t_pk (a int, PRIMARY KEY(a, gp_segment_id)) distributed replicated;
ERROR:  cannot alter system column "gp_segment_id"

nishants4=# ALTER TABLE repli_t_pk ADD CONSTRAINT pk_ctid UNIQUE(ctid);
ERROR:  cannot create constraint or unique index on system column

nishants4=# ALTER TABLE repli_t_pk ADD CONSTRAINT pk_ctid UNIQUE(gp_segment_id);
ERROR:  cannot create constraint or unique index on system column
```

gpopt reviewed (approved):
+1

chrishajas commented on src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:
```diff
@@ -2655,10 +2655,9 @@ CTranslatorRelcacheToDXL::ConstructAttnoMapping(CMemoryPool *mp,
 //
```
Can we also add a comment here about what a keyset is

24nishant replied:
Yes, updated.

DevChattopadhyay reviewed (approved):
Nice Work Nishant. Learned a few things from this PR. The changes look good to me.
